systemd/systemd.go

```diff
@@ -614,14 +623,45 @@ func (s *systemd) getUnitStatus(properties []string, unitNames []string) ([]*Uni
 				}
 			}
 			if len(missing) > 0 {
-				return nil, fmt.Errorf("cannot get unit %q status: missing %s in ‘systemctl show’ output", cur.UnitName, strings.Join(missing, ", "))
+				return nil, fmt.Errorf("cannot get unit %q status: missing %s in ‘systemctl show’ output", cur.Id, strings.Join(missing, ", "))
 			}
 			sts = append(sts, cur)
+
+			// If we received more replies than what we requested, bail.
 			if len(sts) > len(unitNames) {
-				break // wut
+				break
+			}
+
+			// The 'systemctl' command can return the status parameters for a
+			// number of units at the same time. The status output between
+			// units are separated with an empty line. Every parsed status
+			// is appended to the 'sts' array, so the n'th entry matches the
+			// n'th request, as ordered in 'unitNames'. The 'sts' array can
+			// therefore be used as an index to the original request.
+			requestIndex := len(sts) - 1
+
+			// Record which unit 'Name' request produced this status because
+			// if the request was made using an aliased name, we must be
+			// consistent and reply using the same alias. We will check the
+			// validity of the alias below.
+			cur.Name = unitNames[requestIndex]
+
+			// The 'Names' DBUS property of the 'Unit' interface exposes all
+			// the unit name aliases as well as the real name 'Id'. In
+			// order to verify if the request matches the reply, we should compare
+			// the request name 'Name' against the 'Names' list in case the
+			// request was made using a name alias (e.g. ctrl-alt-del.target).
+			// The 'Names' unit property exist for all derived 'Unit' types,
+			// including services, targets, sockets, mounts and timers.
+			validReply := false
+			for _, name := range cur.Names {
+				if cur.Name == name {
+					validReply = true
+					break
+				}
 			}
-			if cur.UnitName != unitNames[len(sts)-1] {
-				return nil, fmt.Errorf("cannot get unit status: queried status of %q but got status of %q", unitNames[len(sts)-1], cur.UnitName)
+			if !validReply {
```

Comment:
this can be done `!(cur.Name == cur.Id || strutil.Contains(cur.Names, cur.Name))` instead of the explicit for loop importing github.com/snapcore/snapd/strutil

Reply:
Fixed. Replaced the loop with strutil.ListContains().

Question: The 'Names' list returned from DBUS explicitly set the Id field to be the first in the list. In other words, the proposed short-circuit evaluation in the proposed line above may not really save much. Would you prefer I still explicitly compare the common case first, or would you be happy with the ListContains() knowing the common case is guaranteed the first element in 'Names' ?

Reply:
I would slightly prefer for us to do the == check as well,  I'm always a bit paranoid with systemd changing its behavior

Reply:
Fixed.